README.md
# GTIRB Stack Stamp

This repository holds example implementations of binary
transformations implemented over top of
[GTIRB](https://github.com/grammatech/gtirb).  See the accompanying
[GTIRB
Tutorial](https://grammatech.github.io/gtirb/md_stack-stamp.html) for
more information.

Specifically, this example repository implements a transform to apply
'stack stamping' protections to a binary.

![stack-stamp signature graphic.](.stack-stamp.svg)

## Abstract

Stack stamping is a technique to help mitigate
<abbr title="Return Oriented Programming">ROP</abbr> style attacks.
This is done by 'stamping' the return address on the stack, thus
encrypting it.  Before it is popped off the stack and used, it is
decrypted by 'un-stamping' it.  This can be an efficient protection,
as no registers are needed, and while flags are affected, they are
only affected at function entry/exits where they do not need to be
preserved.  By encoding and decoding this return address, an attacker
has a more difficult task, since the replacement data would need to be
properly encoded, such that when it is un-stamped, it results in the
desired address.

## Building

This repository contains three implementations of stack-stamping in three
different languages:

1. [Python](#python)
2. [C++](#c)
3. [Common Lisp](#common-lisp)

### Python

The Python transform requires some dependencies to be installed:

```sh
pip3 install gtirb-capstone gtirb-functions capstone keystone-engine
```

To starting using it, run:

```sh
python3 setup.py develop
```

To invoke the command line utility thus generated:

```sh
python3 -m gtirb_stack_stamp
```

### C++

This transform depends on the following libraries:

* [Boost](https://www.boost.org/) (version 1.67 or later)
* [GTIRB](https://github.com/grammatech/gtirb)
* [Capstone](https://github.com/aquynh/capstone)
* [Keystone](https://github.com/keystone-engine/keystone)

Ensure they are installed before compiling the C++ version of the transform.
Building from source also depends on [CMake](https://cmake.org) being installed.

#### Options

We add the filling CMake options during building:

* `GTIRB_STACK_STAMP_ENABLE_TESTS`: Set to `OFF` to disable the downloading of
  Google Test and the building of the test executable. `ON` by default.
* `GTIRB_STACK_STAMP_BUILD_SHARED_LIBS`: Set to `OFF` to build static libraries
  instead of dybnamic ones. `ON` by default.

#### On Linux

```sh
cmake -Bbuild ./
cd build
make
```

The generated command-line utility will then be available in `build/bin`.

#### On Windows

Currently, [some](https://github.com/keystone-engine/keystone/issues/471)
[issues](https://github.com/keystone-engine/keystone/issues/472) are preventing
Keystone from being built on Windows, so the C++ version of gtirb-stack-stamp is
buildable on Linux only for the time being.

#### Tests

Our CMake automatically downloads a local copy of [Google Test](https://github.com/google/googletest)
and produces a test executable. To run it:

```sh
build/bin/test-gtirb-stack-stamp
```

You will need [gtirb-pprinter](https://github.com/grammatech/gtirb-pprinter)
and [ddisasm](https://github.com/grammatech/ddisasm) on your PATH.

### Common Lisp

The Common Lisp transform requires the following external libraries:
- [Capstone](https://github.com/aquynh/capstone)
- [Keystone](https://github.com/keystone-engine/keystone)

and the following common lisp packages
([gtirb](https://github.com/grammatech/gtirb),
 [gtirb-functions](https://github.com/grammatech/gtirb-functions),
 [gtirb-capstone](https://github.com/grammatech/gtirb-capstone))
which may be installed via QuickLisp:

1. Clone this repository into your `~/quicklisp/local-projects` directory
    ```sh
    git clone https://github.com/grammatech/gtirb-stack-stamp
    ```

2. Load `gtirb-stack-stamp` and all dependencies.
    ```lisp
    (ql:quickload :gtirb-stack-stamp)
    ```

To run the transform at the REPL:
```lisp
(write-gtirb (stack-stamp (drop-cfi (read-gtirb "in.gtirb"))) "out.gtirb")
```

To build the command line executable:
```sh
sbcl --eval '(ql:quickload :gtirb-stack-stamp)' \
     --eval '(asdf:make :gtirb-stack-stamp :type :program :monolithic t)'
```

To invoke the command line utility thus generated:
```sh
./stack-stamp --help
```

## Copyright and Acknowledgments

Copyright (C) 2020 GrammaTech, Inc.

This code is licensed under the MIT license. See the LICENSE file in
the project root for license terms.

This project is sponsored by the Office of Naval Research, One Liberty
Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
N68335-17-C-0700.  The content of the information does not necessarily
reflect the position or policy of the Government and no official
endorsement should be inferred.


CHANGELOG.md
0.1.1: (unreleased)

 - TBD

0.1.0:

 - Initial release


gtirb-stack-stamp.lisp
;;; Copyright (C) 2020 GrammaTech, Inc.
;;;
;;; This code is licensed under the MIT license. See the LICENSE file in
;;; the project root for license terms.
;;;
;;; This project is sponsored by the Office of Naval Research, One Liberty
;;; Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
;;; N68335-17-C-0700.  The content of the information does not necessarily
;;; reflect the position or policy of the Government and no official
;;; endorsement should be inferred.
(defpackage :gtirb-stack-stamp/gtirb-stack-stamp
  (:nicknames :gtirb-stack-stamp)
  (:use :gt/full :gtirb :gtirb-functions :gtirb-capstone
        :stefil :command-line-arguments)
  (:shadowing-import-from :gt :size)
  (:import-from :cl-intbytes :int->octets :octets->uint)
  (:import-from :asdf/system :system-relative-pathname)
  (:shadow :version :architecture :mode :symbol)
  (:export :gtirb-stack-stamp))
(in-package :gtirb-stack-stamp/gtirb-stack-stamp)
(in-readtable :curry-compose-reader-macros)
(defmethod size ((obj gtirb-node)) (gtirb:size obj))

;;; Implementation
(defgeneric stack-stamp (object)
  (:documentation "Apply the stack-stamp transformation.")
  (:method ((obj gtirb))
    (set-syntax obj :syntax-att)
    (mapc #'stack-stamp (modules obj)))
  (:method ((obj module)) (mapc #'stack-stamp (functions obj)))
  (:method ((obj func))
    (unless (entries obj) (return-from stack-stamp))
    (let ((exits (append (returns obj) (tail-calls obj))))
      (unless exits (return-from stack-stamp))
      (let* ((key (int->octets (sxhash obj) 8))
             (stamp-bytes (asm obj (format nil "xorl $0x~x,(%rsp); ~
                                                xorl $0x~x,4(%rsp);"
                                           (octets->uint (subseq key 0 4) 4)
                                           (octets->uint (subseq key 4) 4)))))
        #+debug (format t "ENTRANCE: 0x~x~%" (address (first (entries obj))))
        (mapc (lambda (entry-block)
                (setf (bytes entry-block 0 0) stamp-bytes))
              (entries obj))
        (mapc (lambda (exit-block)
                (if-let* ((ins (instructions exit-block))
                          ;; Ret instruction *or* last instruction (tail call).
                          (ins-pos (or (position-if [{eql :ret} #'mnemonic] ins)
                                       (1- (length ins))))
                          ;; Convert instruction index into byte index.
                          (byte-pos (reduce #'+ (subseq ins 0 ins-pos)
                                            :key #'size)))
                  (progn
                    #+debug (format t "EXIT: 0x~x~%" (address exit-block))
                    (setf (bytes exit-block byte-pos byte-pos) stamp-bytes))
                  (error "Unable to find stamp return block for ~s" obj)))
              exits)))))

(defmethod stack-stamp :around ((obj gtirb-node)) (call-next-method) obj)


;;;; Main test suite.
(defsuite test)
(in-suite test)

(defvar *hello*)

(defixture hello
  (:setup (setf *hello* (read-gtirb
                         (system-relative-pathname "gtirb-stack-stamp"
                                                   "tests/hello.v1.gtirb"))))
  (:teardown (setf *hello* nil)))

(defun drop-cfi (ir)
  (mapc (lambda (module)
          (setf (aux-data module)
                (remove-if [{string= "cfiDirectives"} #'car]
                           (aux-data module))))
        (modules ir))
  ir)

(deftest stack-stamp-hello ()
  (nest
   (with-fixture hello)
   (flet ((symbolic-expressions (it)
            (nest (mappend [#'hash-table-values #'symbolic-expressions])
                  (mappend #'byte-intervals) (mappend #'sections)
                  (modules it)))
          (interval-bytes (it)
            (nest (apply #'concatenate 'vector)
                  (mapcar #'bytes)
                  (mappend #'byte-intervals) (mappend #'sections)
                  (modules it)))
          (block-bytes (it)
            (nest (apply #'concatenate 'vector)
                  (mapcar #'bytes) (mappend #'blocks)
                  (mappend #'byte-intervals) (mappend #'sections)
                  (modules it)))))
   (let ((original-symbolic-expressions (symbolic-expressions *hello*))
         (original-interval-bytes (interval-bytes *hello*))
         (original-bytes (block-bytes *hello*)))
     (is (typep (stack-stamp *hello*) 'gtirb))
     (let ((new-symbolic-expressions (symbolic-expressions *hello*)))
       ;; Should not have fewer symbolic expressions when we're only
       ;; inserting (not removing) instructions (and therefore bytes).
       (is (= (length original-symbolic-expressions)
              (length new-symbolic-expressions)))
       ;; In fact we should have exactly the same instructions.
       (is (= (length original-symbolic-expressions)
              (length new-symbolic-expressions))))
     (let ((new-bytes (block-bytes *hello*)))
       ;; We should have more bytes (with the new stamping instructions).
       (is (< (length original-interval-bytes)
              (length (interval-bytes *hello*))))
       (is (< (length original-bytes) (length new-bytes)))))))


;;;; External command-line driver.
;;;
;;; Compile with the following command:
;;;   sbcl --eval '(ql:quickload :gtirb-stack-stamp)' \
;;;        --eval '(asdf:make :gtirb-stack-stamp :type :program :monolithic t)'
;;;
(define-command ss
    (input output
           &spec
           '((("help" #\h #\?) :type boolean :optional t
              :documentation "display help output")
             (("gtirb" #\g) :type boolean :optional t :initial-value nil
              :documentation "output binary gtirb (default)")
             (("asm" #\a) :type boolean :optional t :initial-value nil
              :documentation "output assembly text")
             (("binary" #\b) :type boolean :optional t :initial-value nil
              :documentation "output a binary executable"))
           &aux ir)
  "Apply \"stack stamp\" protections to a binary executable." ""
  (when help (show-help-for-ss))
  (unless (or gtirb asm binary)
    (error "Must specify at least one output type: gtirb, asm, or binary."))
  (setf ir
        ;; If INPUT is a path that doesn't end in "gtirb" call ddisasm.
        (if (and (or (pathnamep input) (stringp input))
                 (not (string= "gtirb" (pathname-type input))))
            (with-temporary-file (:pathname temp :type "gtirb" :direction input)
              (wait-process
               (launch-program
                (list "ddisasm" (namestring input) "--ir" (namestring temp))
                :output *standard-output*
                :error-output *error-output*))
              (read-gtirb temp))
            ;; Otherwise we assume our input is already a GTIRB file.
            (read-gtirb input)))
  (setf ir (stack-stamp (drop-cfi ir)))
  (labels ((normalize-path (others extension)
             (let ((new (if others
                            (make-pathname :type extension :defaults output)
                            output)))
               (when (and (not (equalp new output)) (probe-file new))
                 (error "Output ~a already exists" (namestring new)))
               (namestring new)))
           (gtirb-path () (normalize-path (or asm binary) "gtirb"))
           (binary-path () (normalize-path (or asm gtirb) nil))
           (asm-path () (normalize-path (or gtirb binary) "s")))
    (with-temporary-file (:pathname temp :type "gtirb" :direction output)
      (write-gtirb ir temp)
      (when gtirb
        (wait-process
         (launch-program (list "cp" (namestring temp) (gtirb-path)))))
      (when (or asm binary)
        (wait-process
         (launch-program
          `("gtirb-pprinter" ,(namestring temp)
                             "--skip-section"".eh_frame"
                             ,@(when binary (list "--binary" (binary-path)))
                             ,@(when asm (list "--asm" (asm-path))))
          :output *standard-output*
          :error-output *error-output*))))))
